### PR TITLE
Add `findUp.exists()` and `findUp.sync.exists()`

### LINKED ISSUE #37: Add utility methods to check if the path exists
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/find-up/issues/37)

This will be useful for #28.

I'm thinking `findUp.exists()` and `findUp.sync.exists()`.

We already depend on https://github.com/sindresorhus/path-exists further down in the dependency tree, so we can just depend on that here without any added weight.

Should be tested and added to the readme and TS definition.

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/find-up/issues/37).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/find-up/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by TiagoDanin**
Closes #37
- [x] Documentation
- [x] TypeScript Definition
- [x] Utility Methods

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+	}
+
+	return fs.existsSync(path);
+};
+
+const stats = path => fs.statSync(path);
+
+const isFile = path => exists(path) ? stats(path).isFile() : false;
+
+const isDirectory = path => exists(path) ? stats(path).isDirectory() : false;
```
Comment (sholladay) [inline index.js]: These functions are vulnerable to race conditions. Use the [path-type](https://github.com/sindresorhus/path-type) module to avoid it.

File: index.d.ts [hunk 2]
```diff
@@ -70,6 +77,30 @@ declare const findUp: {
 	/**
 	Return this in a `matcher` function to stop the search and force `findUp` to immediately return `undefined`.
 	*/
+	exists(path: string): boolean;
```
Comment (coreyfarrell) [inline index.d.ts]: I think having `findUp.exists` be a sync function is inconsistent with the rest of the module, everything else is async first so `findUp.exists` should return `Promise<boolean>`.  This comment applies to all functions being added.

I don't know if the sync version should be `findUp.sync.exists` or `findUp.exists.sync`.  I lean towards `findUp.sync.exists`, this way someone could use:
```js
const findUp = require('find-up').sync;
findUp.exists('some-file.js');
```

File: index.js (hunk 1 above)
Comment (coreyfarrell) [reply index.js]: I commented at https://github.com/sindresorhus/find-up/issues/37#issuecomment-489440312, maybe `locate-path` should be used for isFile and isDirectory functionality?  Not sure if this means that `locate-path` should simply be re-exported instead of these helpers or if the helpers should use `locate-path` for the implementation.

Obviously this idea is dependent on updating `locate-path` first.  I'm not really asking for this change to happen here yet, just putting the idea out there.

**Comment by TiagoDanin** [comment]
`index.d.ts`
```ts
declare const findUp: {
	// ... //
	sync.exists(path: string): boolean; //This does not create a nested function (?)
	// ... //
}
```

File: index.d.ts (hunk 2 above)
Comment (sindresorhus) [reply index.d.ts]: > I lean towards findUp.sync.exists, this way someone could use:

👍

File: index.d.ts [hunk 3]
```diff
@@ -70,7 +77,41 @@ declare const findUp: {
...
+	/**
+	Returns `Promise` when resolved if `true` the path is file.
+
+	@param path - Path to the file or directory.
+	@returns `Promise<boolean>`.
+	*/
+	isSymlink(path: string): Promise<boolean>;
+
+	/**
+	Returns `Promise` when resolved if `true` the path is symlink.
```
Comment (sindresorhus) [inline index.d.ts]: The docs comments are not for the correct methods.

**Comment by TiagoDanin** [comment]
@coreyfarrell I believe this is a hard link [[Source](https://www.cyberciti.biz/faq/unix-creating-symbolic-link-ln-command/)]
https://github.com/sindresorhus/find-up/blob/da1b331f95f503fb12057a4b040586c1f0bda77c/test.js#L40

@sindresorhus create `link.js` with `ln -s fixture/baz.js fixture/link.js` and run `pathExists('fixture/link.js') // false`, is that right ?

**Comment by coreyfarrell** [comment]
@TiagoDanin it might be a hard link in Windows.  The symbolic link tests need to be skipped under Windows (see other tests which try to match the links).

**Comment by sindresorhus** [comment]
Because of https://github.com/sindresorhus/find-up/pull/42, we no longer need the isFile/isDirectory methods. We only need the "exists" methods.

File: test.js [hunk 4]
```diff
@@ -50,6 +50,8 @@ test.afterEach(t => {
 	fs.rmdirSync(t.context.disjoint);
 });
 
+const isNotWindows = process.platform !== 'win32';
```
Comment (sindresorhus) [inline test.js]:
```suggestion
const isWindows = process.platform === 'win32';
```

And use it with `!`.

File: index.d.ts [hunk 5]
```diff
@@ -50,26 +57,44 @@ declare const findUp: {
...
+	}
 
 	/**
-	Synchronously find a file or directory by walking up parent directories.
+	Returns `Promise` when resolved if `true` the path is exists.
 
-	@param matcher - Called for each directory in the search. Return a path or `findUp.stop` to stop the search.
-	@returns The first path found or `undefined` if none could be found.
+	@param path - Path to the file or directory.
+	@returns `Promise<boolean>`.
```
Comment (sindresorhus) [inline index.d.ts]: https://github.com/sindresorhus/typescript-definition-style-guide#documentation

```
@returns should not duplicate the type information unless it's impossible to describe it without.
@returns A boolean of whether it was enabled. → @returns Whether it was enabled.
```

File: index.d.ts [hunk 6]
```diff
@@ -50,26 +57,44 @@ declare const findUp: {
...
+	exists(path: string): Promise<boolean>;
 
 	/**
 	Return this in a `matcher` function to stop the search and force `findUp` to immediately return `undefined`.
 	*/
 	readonly stop: findUp.StopSymbol;
 };
 
+
+
```
Comment (sindresorhus) [inline index.d.ts]: Don't do unrelated changes.

Comment (TiagoDanin) [reply index.d.ts]: Ops!

**Comment by sindresorhus** [comment]
@coreyfarrell Looks good to you?

File: index.d.ts [hunk 7]
```diff
@@ -37,6 +38,12 @@ declare const findUp: {
 
 		console.log(await findUp(['rainbow.png', 'unicorn.png']));
 		//=> '/Users/sindresorhus/unicorn.png'
+
+		console.log(await findUp(async directory => {
```
Comment (coreyfarrell) [inline index.d.ts]: Should this example be moved into the comment for the matcher variant?

File: readme.md [hunk 8]
```diff
@@ -28,18 +28,17 @@ $ npm install find-up
 ```
 /
 └── Users
-		└── sindresorhus
-				├── unicorn.png
-				└── foo
-						└── bar
-								├── baz
-								└── example.js
+	└── sindresorhus
```
Comment (coreyfarrell) [inline readme.md]: This is an improvement but I feel like space for indent would be better than tab.

File: index.d.ts (hunk 7 above)
Comment (sindresorhus) [reply index.d.ts]: Yes

File: index.d.ts [hunk 9]
```diff
@@ -50,21 +57,37 @@ declare const findUp: {
...
-	/**
-	Synchronously find a file or directory by walking up parent directories.
-
-	@param name - Name of the file or directory to find. Can be multiple.
-	@returns The first path found (by respecting the order of `name`s) or `undefined` if none could be found.
-	*/
-	sync(name: string | string[], options?: findUp.Options): string | undefined;
+	sync: {
+		/**
+		Synchronously find a file or directory by walking up parent directories.
```
Comment (sindresorhus) [inline index.d.ts]: Empty line after description.

File: index.d.ts [hunk 10]
```diff
@@ -21,14 +21,15 @@ declare const findUp: {
...
-	// 								└── example.js
+	//     └── sindresorhus
+	//         ├── unicorn.png
+	//         └── foo
+	//             └── bar
+	//                 ├── baz
+	//                 └── example.js
 
 	// example.js
+	import path = require('path');
```
Comment (sindresorhus) [inline index.d.ts]: ? This example is not using `path`.
```suggestion
```
